File: src/creatioClient.ts
/**
 * Shared, read-only Creatio OData client.
 *
 * SAFETY MODEL
 *  - Every network call goes through odataGet(), whose HTTP method is hard-coded
 *    to GET. There is no create/update/delete path anywhere in this module, so
 *    every consumer (MCP server, web app) is read-only by construction.
 *  - An optional entity allowlist (CREATIO_ALLOWED_ENTITIES) restricts which
 *    OData entity sets are reachable; $top is clamped to CREATIO_MAX_TOP.
 *
 * AUTH — two modes, auto-selected:
 *  1. Forms auth via /ServiceModel/AuthService.svc/Login (CREATIO_LOGIN +
 *     CREATIO_PASSWORD). Session re-established automatically on 401.
 *  2. Cookie auth (CREATIO_ASPXAUTH + CREATIO_BPMCSRF [+ CREATIO_BPMLOADER]) for
 *     SSO tenants. The cookies live in the .env FILE and are re-read on demand,
 *     so refreshing them there takes effect on the next query — no restart.
 */

import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { readFileSync, writeFileSync } from "node:fs";

// .env sits next to the project root (../.env relative to dist/creatioClient.js).
export const ENV_PATH = join(dirname(fileURLToPath(import.meta.url)), "..", ".env");

// Load .env into process.env for stable config (base URL, allowlist, row cap).
// NOTE: loadEnvFile does NOT override vars already in process.env, so anything
// the MCP client baked into its registration wins here — fine for stable config.
// SSO cookies are handled separately (resolveCookieEnv), reading the file live.
try {
  (process as { loadEnvFile?: (p: string) => void }).loadEnvFile?.(ENV_PATH);
} catch {
  // No .env next to the module — fall back to whatever is already in process.env.
}

/**
 * Parse KEY=VALUE lines straight from the .env file on disk. Returns {} if the
 * file can't be read. Called live (not cached) so a cookie refresh in .env is
 * picked up on the next request without restarting the process.
 */
export function readEnvFile(): Record<string, string> {
  try {
    const out: Record<string, string> = {};
    for (const raw of readFileSync(ENV_PATH, "utf8").split(/\r?\n/)) {
      const line = raw.trim();
      if (!line || line.startsWith("#")) continue;
      const eq = line.indexOf("=");
      if (eq === -1) continue;
      const key = line.slice(0, eq).trim();
      let val = line.slice(eq + 1).trim();
      if (
        (val.startsWith('"') && val.endsWith('"')) ||
        (val.startsWith("'") && val.endsWith("'"))
      ) {
        val = val.slice(1, -1);
      }
      out[key] = val;
    }
    return out;
  } catch {
    return {};
  }
}

/**
 * Merge-write keys into the .env file, preserving existing comments, ordering,
 * and untouched keys. Keys present in `updates` are replaced in place; new keys
 * are appended. Used by the web app's Settings screen.
 */
export function writeEnvFile(updates: Record<string, string>): void {
  let lines: string[] = [];
  try {
    lines = readFileSync(ENV_PATH, "utf8").split(/\r?\n/);
  } catch {
    lines = [];
  }
  const remaining = new Map(Object.entries(updates));
  const out = lines.map((line) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) return line;
    const eq = trimmed.indexOf("=");
    if (eq === -1) return line;
    const key = trimmed.slice(0, eq).trim();
    if (remaining.has(key)) {
      const val = remaining.get(key)!;
      remaining.delete(key);
      return `${key}=${val}`;
    }
    return line;
  });
  for (const [key, val] of remaining) out.push(`${key}=${val}`);
  // Trailing newline, no duplicate blank lines at EOF.
  while (out.length && out[out.length - 1] === "") out.pop();
  writeFileSync(ENV_PATH, out.join("\n") + "\n", "utf8");
}

// ---------------------------------------------------------------------------
// Config resolution
// ---------------------------------------------------------------------------
function envFirst(name: string): string {
  const v = process.env[name];
  if (v !== undefined && v !== "") return v.trim();
  return (readEnvFile()[name] ?? "").trim();
}

function clampInt(raw: string | undefined, dflt: number, min: number, max: number): number {
  const n = raw ? parseInt(raw, 10) : NaN;
  if (Number.isNaN(n)) return dflt;
  return Math.max(min, Math.min(max, n));
}

// Stable config (resolved once at import; changing these needs a restart).
export const BASE_URL = envFirst("CREATIO_BASE_URL").replace(/\/+$/, "");
export const ALLOWED_ENTITIES = envFirst("CREATIO_ALLOWED_ENTITIES")
  .split(",")
  .map((s) => s.trim())
  .filter(Boolean);
export const MAX_TOP = clampInt(envFirst("CREATIO_MAX_TOP") || undefined, 50, 1, 500);

// FileService entities the download proxy may fetch (case/feed/email attachments).
// Kept tight so the proxy can't be pointed at arbitrary file records.
export const FILE_DOWNLOAD_ENTITIES = ["CaseFile", "FeedFile", "ActivityFile"];

/**
 * Resolve the three SSO cookies, preferring the .env FILE on disk over
 * process.env — so refreshing .env (and nothing else) always takes effect, even
 * if the MCP client's registration baked in stale cookie values. Normalization
 * mirrors the accepted paste formats (bare value or `Name=value`).
 */
export function resolveCookieEnv(): { aspx: string; csrf: string; loader: string } {
  const file = readEnvFile();
  const pick = (name: string) => (file[name] ?? process.env[name] ?? "").trim();
  return {
    aspx: pick("CREATIO_ASPXAUTH").replace(/^\.?ASPXAUTH=/i, ""),
    csrf: pick("CREATIO_BPMCSRF").replace(/^BPMCSRF=/i, ""),
    loader: pick("CREATIO_BPMLOADER").replace(/^BPMLOADER=/i, ""),
  };
}

const INITIAL_COOKIES = resolveCookieEnv();
export const COOKIE_MODE = Boolean(INITIAL_COOKIES.aspx && INITIAL_COOKIES.csrf);

// Forms auth is only used when NOT in cookie mode.
const LOGIN = envFirst("CREATIO_LOGIN");
const PASSWORD = envFirst("CREATIO_PASSWORD");

export function assertEntityAllowed(entity: string): void {
  if (!/^[A-Za-z][A-Za-z0-9_]*$/.test(entity)) {
    throw new Error(`Invalid entity name: ${entity}`);
  }
  if (ALLOWED_ENTITIES.length > 0 && !ALLOWED_ENTITIES.includes(entity)) {
    throw new Error(
      `Entity "${entity}" is not in the allowlist (${ALLOWED_ENTITIES.join(", ")}). ` +
        `Edit CREATIO_ALLOWED_ENTITIES to permit it.`
    );
  }
}

// ---------------------------------------------------------------------------
// Auth + HTTP (GET only)
// ---------------------------------------------------------------------------
let cookieJar = new Map<string, string>();
let bpmcsrf = "";
// What seedCookiesFromEnv() last read from disk — used to detect a .env refresh.
let lastSeededAspx = "";
let lastSeededCsrf = "";

function cookieHeader(): string {
  return [...cookieJar.entries()].map(([k, v]) => `${k}=${v}`).join("; ");
}

function storeSetCookies(res: Response): void {
  // Node 18.14+ / undici: getSetCookie() returns all Set-Cookie values.
  const raw = (res.headers as any).getSetCookie?.() as string[] | undefined;
  const list = raw ?? (res.headers.get("set-cookie") ? [res.headers.get("set-cookie")!] : []);
  for (const line of list) {
    const first = line.split(";")[0];
    const eq = first.indexOf("=");
    if (eq === -1) continue;
    const name = first.slice(0, eq).trim();
    const value = first.slice(eq + 1).trim();
    cookieJar.set(name, value);
    if (name === "BPMCSRF") bpmcsrf = value;
  }
}

/** Seed the cookie jar from the .env file (SSO cookie mode), reading the
 *  freshest values from disk on every call so a refresh is picked up live. */
export function seedCookiesFromEnv(): void {
  const { aspx, csrf, loader } = resolveCookieEnv();
  cookieJar = new Map();
  cookieJar.set(".ASPXAUTH", aspx);
  cookieJar.set("BPMCSRF", csrf);
  if (loader) cookieJar.set("BPMLOADER", loader);
  bpmcsrf = csrf;
  lastSeededAspx = aspx;
  lastSeededCsrf = csrf;
}

/** Thrown when SSO cookies are missing/expired. Consumers can special-case it
 *  (the web app maps it to an "open Settings" banner). */
export class AuthError extends Error {}

export async function login(): Promise<void> {
  if (!BASE_URL) {
    throw new Error("Creatio base URL is not configured (CREATIO_BASE_URL). Set it in Settings / .env.");
  }

  // Cookie mode can't re-authenticate against Creatio (the cookies came from a
  // browser session), but the cookies live in .env — so on expiry we re-read the
  // file: if you've refreshed it, we pick up the new cookies and carry on; if
  // it's unchanged, we report the clear error. No restart needed either way.
  if (COOKIE_MODE) {
    const fresh = resolveCookieEnv();
    const changed = fresh.aspx !== lastSeededAspx || fresh.csrf !== lastSeededCsrf;
    if (bpmcsrf && !changed) {
      throw new AuthError(
        "Creatio session cookies have expired. Grab fresh .ASPXAUTH / BPMCSRF / BPMLOADER " +
          "from your browser (DevTools → Application → Cookies) and update your .env. " +
          "The server re-reads .env automatically — no restart needed."
      );
    }
    seedCookiesFromEnv();
    return;
  }

  cookieJar = new Map();
  bpmcsrf = "";
  const res = await fetch(`${BASE_URL}/ServiceModel/AuthService.svc/Login`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify({ UserName: LOGIN, UserPassword: PASSWORD }),
  });
  storeSetCookies(res);
  const body = (await res.json().catch(() => ({}))) as { Code?: number; Message?: string };
  if (!res.ok || (body.Code !== undefined && body.Code !== 0)) {
    throw new Error(`Creatio login failed (Code=${body.Code}): ${body.Message ?? res.statusText}`);
  }
  if (!bpmcsrf) throw new Error("Creatio login succeeded but no BPMCSRF cookie was returned.");
}

/**
 * Flatten Creatio's nested OData error envelope into one readable line.
 *
 * The raw body is
 *   {"error":{"message":"An error has occurred.","innererror":{"message":"...",
 *    "internalexception":{"message":"<the actual cause>"}}}}
 * — the generic outer text is useless, and the real reason sits at the bottom.
 * Walk to the innermost non-empty message; fall back to the raw text if the
 * body isn't the shape we expect (HTML error page, empty 500, etc.).
 */
function odataErrorMessage(text: string): string {
  let msg = "";
  try {
    let node = JSON.parse(text)?.error;
    while (node && typeof node === "object") {
      if (typeof node.message === "string" && node.message.trim()) msg = node.message.trim();
      node = node.innererror ?? node.internalexception;
    }
  } catch {
    // Not JSON — fall through to the raw body.
  }
  if (!msg) return text.slice(0, 500) || "(empty response body)";

  // The all-columns serialization failure (Creatio's published EDM model listing
  // a column the runtime record doesn't have) is the one 500 we hit repeatedly,
  // and the fix is always the same — say so instead of making the next person
  // rediscover it. See districtIndex.ts CASE_EXPAND for the $expand variant.
  if (/failed to serialize the response body|missing the property/i.test(msg)) {
    msg += " — this Creatio entity cannot be read with all columns; pass an explicit $select.";
  }
  return msg;
}

/** The ONLY network call path. Method is hard-coded to GET. */
export async function odataGet(path: string): Promise<any> {
  if (!bpmcsrf) await login();

  const doFetch = () =>
    fetch(`${BASE_URL}/0/odata/${path}`, {
      method: "GET",
      headers: {
        Accept: "application/json",
        Cookie: cookieHeader(),
        BPMCSRF: bpmcsrf,
        ForceUseSession: "true",
      },
    });

  let res = await doFetch();
  if (res.status === 401 || res.status === 403) {
    await login(); // session expired — re-auth (or re-read refreshed cookies) once and retry
    res = await doFetch();
  }

  const text = await res.text();
  if (res.status === 401 || res.status === 403) {
    throw new AuthError(
      "Creatio rejected the session (HTTP " +
        res.status +
        "). The cookies are expired or invalid — update them in Settings / .env."
    );
  }
  if (!res.ok) {
    // No res.statusText: undici leaves it empty on HTTP/2 (no reason phrase).
    throw new Error(`OData GET /${path} -> ${res.status}: ${odataErrorMessage(text)}`);
  }
  return text ? JSON.parse(text) : {};
}

/** Download an attachment (read-only GET) from Creatio's FileService, e.g.
 *  /0/rest/FileService/Download/CaseFile/<guid>. Restricted to the file-entity
 *  allowlist + a GUID id so it can't be aimed at arbitrary paths. */
export async function downloadFile(
  entity: string,
  id: string
): Promise<{ buffer: Buffer; contentType: string; filename?: string }> {
  if (!FILE_DOWNLOAD_ENTITIES.includes(entity)) {
    throw new Error(`File entity "${entity}" is not downloadable.`);
  }
  if (!/^[0-9a-fA-F-]{36}$/.test(id)) throw new Error(`Invalid file id: ${id}`);
  if (!BASE_URL) throw new Error("Creatio base URL is not configured.");
  if (!bpmcsrf) await login();

  const url = `${BASE_URL}/0/rest/FileService/Download/${entity}/${id}`;
  const doFetch = () =>
    fetch(url, {
      method: "GET",
      headers: { Accept: "*/*", Cookie: cookieHeader(), BPMCSRF: bpmcsrf, ForceUseSession: "true" },
    });

  let res = await doFetch();
  if (res.status === 401 || res.status === 403) {
    await login();
    res = await doFetch();
  }
  if (res.status === 401 || res.status === 403) {
    throw new AuthError(`Creatio rejected the session (HTTP ${res.status}) while downloading a file.`);
  }
  if (!res.ok) {
    throw new Error(`File download ${entity}/${id} -> ${res.status} ${res.statusText}`);
  }
  const contentType = res.headers.get("content-type") || "application/octet-stream";
  const cd = res.headers.get("content-disposition") || "";
  const filename = (cd.match(/filename\*?=(?:UTF-8'')?"?([^";]+)"?/i) || [])[1];
  const buffer = Buffer.from(await res.arrayBuffer());
  return { buffer, contentType, filename };
}

// ---------------------------------------------------------------------------
// OData query builder (read-only)
// ---------------------------------------------------------------------------
export function buildQuery(opts: {
  select?: string[];
  filter?: string;
  orderby?: string;
  top?: number;
  expand?: string;
}): string {
  const p = new URLSearchParams();
  if (opts.select?.length) p.set("$select", opts.select.join(","));
  if (opts.filter) p.set("$filter", opts.filter);
  if (opts.orderby) p.set("$orderby", opts.orderby);
  if (opts.expand) p.set("$expand", opts.expand);
  p.set("$top", String(clampInt(String(opts.top ?? MAX_TOP), MAX_TOP, 1, MAX_TOP)));
  const s = p.toString();
  return s ? `?${s}` : "";
}

/** Convenience: query an entity set with the standard options. Returns the
 *  OData `value` array (or [] if absent). */
export async function queryRecords(
  entity: string,
  opts: { select?: string[]; filter?: string; orderby?: string; top?: number; expand?: string } = {}
): Promise<any[]> {
  assertEntityAllowed(entity);
  const data = await odataGet(entity + buildQuery(opts));
  return (data && data.value) || [];
}

// ---------------------------------------------------------------------------
// Bulk paging (app-authored queries only)
// ---------------------------------------------------------------------------
// MAX_TOP exists to stop a *model-driven* query from dragging back the whole
// table, and buildQuery() enforces it for every MCP/UI-facing path. Bulk
// indexing has the opposite need: it is app-authored, the page size is a
// constant in our own source, and 50-row pages would turn a 185-request build
// into a 3,700-request one. So it gets its own builder with a higher ceiling
// rather than loosening the guardrail everyone else goes through.
//
// This is still GET-only via odataGet(), so the read-only-by-construction
// property of the module is unchanged.

/** Page size ceiling for app-authored bulk reads. Verified server-side: $top=1000
 *  returns 1000 rows in ~1.4s. */
export const MAX_PAGE_TOP = 1000;

export interface PageOptions {
  select?: string[];
  filter?: string;
  orderby?: string;
  top?: number;
  skip?: number;
  expand?: string;
  /** Ask Creatio for the total matching row count (`@odata.count` on the response). */
  count?: boolean;
}

/** buildQuery's bulk sibling: honours $skip/$count and clamps to MAX_PAGE_TOP. */
export function buildPageQuery(opts: PageOptions): string {
  const p = new URLSearchParams();
  if (opts.select?.length) p.set("$select", opts.select.join(","));
  if (opts.filter) p.set("$filter", opts.filter);
  if (opts.orderby) p.set("$orderby", opts.orderby);
  if (opts.expand) p.set("$expand", opts.expand);
  if (opts.skip && opts.skip > 0) p.set("$skip", String(Math.floor(opts.skip)));
  if (opts.count) p.set("$count", "true");
  p.set(
    "$top",
    String(clampInt(String(opts.top ?? MAX_PAGE_TOP), MAX_PAGE_TOP, 1, MAX_PAGE_TOP))
  );
  const s = p.toString();
  return s ? `?${s}` : "";
}

/** One bulk page. Returns the rows plus the server total when `count` was set. */
export async function pageRecords(
  entity: string,
  opts: PageOptions = {}
): Promise<{ rows: any[]; total?: number }> {
  assertEntityAllowed(entity);
  const data = await odataGet(entity + buildPageQuery(opts));
  const total = data && typeof data["@odata.count"] === "number" ? data["@odata.count"] : undefined;
  return { rows: (data && data.value) || [], total };
}

/** Row count only, via `$count=true&$top=1`. Creatio's `/$count` endpoint returns
 *  non-JSON, so it is not usable through odataGet(). */
export async function countRecords(entity: string, filter?: string): Promise<number> {
  const { total } = await pageRecords(entity, { select: ["Id"], filter, top: 1, count: true });
  return total ?? 0;
}

/** Lightweight auth/connectivity check — reads 1 row of the first allowed
 *  entity (or Contact). Returns {ok:true} or {ok:false, error}. */
export async function testConnection(): Promise<{ ok: boolean; error?: string }> {
  const entity = ALLOWED_ENTITIES[0] || "Contact";
  try {
    // $select=Id, never an all-columns read: Creatio 500s serializing entities
    // whose published EDM model advertises a column the runtime record doesn't
    // have (Case/NltAdditionalInformation, and Contact likewise on this tenant),
    // and an all-columns Account read is slow enough to time out. Id exists on
    // every entity, so this probes auth and nothing else.
    await odataGet(entity + buildQuery({ select: ["Id"], top: 1 }));
    return { ok: true };
  } catch (e) {
    return { ok: false, error: e instanceof Error ? e.message : String(e) };
  }
}

File: creatio_case_lookup/creatio_client.py
"""Shared, read-only Creatio OData client.

SAFETY MODEL
 - Every OData call goes through :func:`odata_get`, whose HTTP method is
   hard-coded to GET. There is no create/update/delete path anywhere in this
   module, so every consumer (MCP server, web app) is read-only by
   construction. (The only POST is the forms-auth login itself.)
 - An optional entity allowlist (CREATIO_ALLOWED_ENTITIES) restricts which
   OData entity sets are reachable; ``$top`` is clamped to CREATIO_MAX_TOP.

AUTH — two modes, auto-selected:
 1. Forms auth via /ServiceModel/AuthService.svc/Login (CREATIO_LOGIN +
    CREATIO_PASSWORD). Session re-established automatically on 401.
 2. Cookie auth (CREATIO_ASPXAUTH + CREATIO_BPMCSRF [+ CREATIO_BPMLOADER]) for
    SSO tenants. The cookies live in the `.env` FILE and are re-read on demand,
    so refreshing them there takes effect on the next query — no restart.

The module-level cookie jar below is the single source of truth for session
cookies. httpx never persists cookies here: every request uses a fresh client
and sends the jar explicitly in a ``Cookie`` header.
"""

from __future__ import annotations

import json
import os
import re
from collections.abc import Mapping
from typing import Any
from urllib.parse import quote_plus

import httpx

from .env import clamp_int, env_first, read_env_file

# ---------------------------------------------------------------------------
# Config resolution
# ---------------------------------------------------------------------------
# Stable config (resolved once at import; changing these needs a restart).
# env.py has already copied `.env` into os.environ WITHOUT overriding values the
# MCP client baked into its registration. SSO cookies are handled separately
# (resolve_cookie_env), reading the file live.
BASE_URL: str = re.sub(r"/+$", "", env_first("CREATIO_BASE_URL"))
ALLOWED_ENTITIES: list[str] = [
    s.strip() for s in env_first("CREATIO_ALLOWED_ENTITIES").split(",") if s.strip()
]
MAX_TOP: int = clamp_int(env_first("CREATIO_MAX_TOP") or None, 50, 1, 500)

# FileService entities the download proxy may fetch (case/feed/email attachments).
# Kept tight so the proxy can't be pointed at arbitrary file records.
FILE_DOWNLOAD_ENTITIES: list[str] = ["CaseFile", "FeedFile", "ActivityFile"]

# fetch() has no timeout at all; a generous one here keeps a dead connection
# from hanging a request forever without cutting off slow-but-healthy reads.
_TIMEOUT = httpx.Timeout(120.0, connect=30.0)

_ENTITY_RE = re.compile(r"[A-Za-z][A-Za-z0-9_]*")
_GUID_RE = re.compile(r"[0-9a-fA-F-]{36}")


def resolve_cookie_env() -> dict[str, str]:
    """Resolve the three SSO cookies, preferring the `.env` FILE on disk over
    the process environment — so refreshing `.env` (and nothing else) always
    takes effect, even if the MCP client's registration baked in stale cookie
    values. Normalization mirrors the accepted paste formats (bare value or
    ``Name=value``). Returns ``{"aspx", "csrf", "loader"}``.
    """
    file = read_env_file()

    def pick(name: str) -> str:
        # A key present in the file wins even when its value is empty.
        v = file[name] if name in file else os.environ.get(name, "")
        return (v or "").strip()

    return {
        "aspx": re.sub(r"^\.?ASPXAUTH=", "", pick("CREATIO_ASPXAUTH"), flags=re.I),
        "csrf": re.sub(r"^BPMCSRF=", "", pick("CREATIO_BPMCSRF"), flags=re.I),
        "loader": re.sub(r"^BPMLOADER=", "", pick("CREATIO_BPMLOADER"), flags=re.I),
    }


_INITIAL_COOKIES = resolve_cookie_env()
COOKIE_MODE: bool = bool(_INITIAL_COOKIES["aspx"] and _INITIAL_COOKIES["csrf"])

# Forms auth is only used when NOT in cookie mode.
_LOGIN = env_first("CREATIO_LOGIN")
_PASSWORD = env_first("CREATIO_PASSWORD")


def assert_entity_allowed(entity: str) -> None:
    if not isinstance(entity, str) or not _ENTITY_RE.fullmatch(entity):
        raise ValueError(f"Invalid entity name: {entity}")
    if ALLOWED_ENTITIES and entity not in ALLOWED_ENTITIES:
        raise ValueError(
            f'Entity "{entity}" is not in the allowlist ({", ".join(ALLOWED_ENTITIES)}). '
            "Edit CREATIO_ALLOWED_ENTITIES to permit it."
        )


# ---------------------------------------------------------------------------
# Auth + HTTP (GET only)
# ---------------------------------------------------------------------------
_cookie_jar: dict[str, str] = {}
_bpmcsrf = ""
# What seed_cookies_from_env() last read from disk — used to detect a .env refresh.
_last_seeded_aspx = ""
_last_seeded_csrf = ""


def _client() -> httpx.AsyncClient:
    """A fresh client per request (no cookie persistence inside httpx).

    fetch() follows redirects by default, so this does too. Tests replace this
    factory to inject an ``httpx.MockTransport``.
    """
    return httpx.AsyncClient(follow_redirects=True, timeout=_TIMEOUT)


def _cookie_header() -> str:
    return "; ".join(f"{k}={v}" for k, v in _cookie_jar.items())


def _store_set_cookies(res: httpx.Response) -> None:
    """Store every Set-Cookie header (name=value only) in the module jar."""
    global _bpmcsrf
    for line in res.headers.get_list("set-cookie"):
        first = line.split(";")[0]
        name, sep, value = first.partition("=")
        if not sep:
            continue
        name, value = name.strip(), value.strip()
        _cookie_jar[name] = value
        if name == "BPMCSRF":
            _bpmcsrf = value


# Public alias under the TS name, for callers/tests.
store_set_cookies = _store_set_cookies


def seed_cookies_from_env() -> None:
    """Seed the cookie jar from the `.env` file (SSO cookie mode), reading the
    freshest values from disk on every call so a refresh is picked up live."""
    global _cookie_jar, _bpmcsrf, _last_seeded_aspx, _last_seeded_csrf
    c = resolve_cookie_env()
    _cookie_jar = {".ASPXAUTH": c["aspx"], "BPMCSRF": c["csrf"]}
    if c["loader"]:
        _cookie_jar["BPMLOADER"] = c["loader"]
    _bpmcsrf = c["csrf"]
    _last_seeded_aspx = c["aspx"]
    _last_seeded_csrf = c["csrf"]


class AuthError(Exception):
    """Raised when SSO cookies are missing/expired. Consumers can special-case
    it (the web app maps it to an "open Settings" banner)."""


def _js_str(v: Any) -> str:
    """String(v) the way JS would print a JSON value in a template literal."""
    if v is None:
        return "null"
    if v is True:
        return "true"
    if v is False:
        return "false"
    if isinstance(v, float) and v.is_integer():
        return str(int(v))
    return str(v)


async def login() -> None:
    global _cookie_jar, _bpmcsrf
    if not BASE_URL:
        raise ValueError(
            "Creatio base URL is not configured (CREATIO_BASE_URL). Set it in Settings / .env."
        )

    # Cookie mode can't re-authenticate against Creatio (the cookies came from a
    # browser session), but the cookies live in .env — so on expiry we re-read
    # the file: if you've refreshed it, we pick up the new cookies and carry on;
    # if it's unchanged, we report the clear error. No restart needed either way.
    # Decide cookie-vs-forms mode from what is on disk RIGHT NOW rather than from
    # the import-time COOKIE_MODE. The app's browser login writes cookies into a
    # .env that may have had none when the server started; that has to take
    # effect without a restart, or we would fall through to forms auth (with
    # empty credentials) and fail even though valid cookies are sitting on disk.
    fresh = resolve_cookie_env()
    if fresh["aspx"] and fresh["csrf"]:
        changed = fresh["aspx"] != _last_seeded_aspx or fresh["csrf"] != _last_seeded_csrf
        if _bpmcsrf and not changed:
            raise AuthError(
                "Creatio session cookies have expired. Grab fresh .ASPXAUTH / BPMCSRF / BPMLOADER "
                "from your browser (DevTools → Application → Cookies) and update your .env. "
                "The server re-reads .env automatically — no restart needed."
            )
        seed_cookies_from_env()
        return

    _cookie_jar = {}
    _bpmcsrf = ""
    async with _client() as client:
        res = await client.post(
            f"{BASE_URL}/ServiceModel/AuthService.svc/Login",
            headers={"Content-Type": "application/json", "Accept": "application/json"},
            content=json.dumps({"UserName": _LOGIN, "UserPassword": _PASSWORD}),
        )
    _store_set_cookies(res)
    try:
        body = res.json()
    except Exception:
        body = {}
    if not isinstance(body, dict):
        body = {}
    if not res.is_success or ("Code" in body and body["Code"] != 0):
        code = _js_str(body["Code"]) if "Code" in body else "undefined"
        msg = body.get("Message")
        raise RuntimeError(
            f"Creatio login failed (Code={code}): {msg if msg is not None else res.reason_phrase}"
        )
    if not _bpmcsrf:
        raise RuntimeError("Creatio login succeeded but no BPMCSRF cookie was returned.")


def _session_headers(accept: str) -> dict[str, str]:
    return {
        "Accept": accept,
        "Cookie": _cookie_header(),
        "BPMCSRF": _bpmcsrf,
        "ForceUseSession": "true",
    }


async def odata_get(path: str) -> Any:
    """The ONLY OData call path. Method is hard-coded to GET."""
    if not _bpmcsrf:
        await login()

    url = f"{BASE_URL}/0/odata/{path}"
    async with _client() as client:
        res = await client.get(url, headers=_session_headers("application/json"))
        if res.status_code in (401, 403):
            # Session expired — re-auth (or re-read refreshed cookies) once and retry.
            await login()
            res = await client.get(url, headers=_session_headers("application/json"))

    text = res.text
    if res.status_code in (401, 403):
        raise AuthError(
            f"Creatio rejected the session (HTTP {res.status_code}). "
            "The cookies are expired or invalid — update them in Settings / .env."
        )
    if not res.is_success:
        raise RuntimeError(
            f"OData GET /{path} -> {res.status_code} {res.reason_phrase}: {text[:500]}"
        )
    return json.loads(text) if text else {}


_FILENAME_RE = re.compile(r"""filename\*?=(?:UTF-8'')?"?([^";]+)"?""", re.I)


async def download_file(entity: str, id: str) -> dict[str, Any]:
    """Download an attachment (read-only GET) from Creatio's FileService, e.g.
    /0/rest/FileService/Download/CaseFile/<guid>. Restricted to the file-entity
    allowlist + a GUID id so it can't be aimed at arbitrary paths.

    Returns ``{"buffer": bytes, "contentType": str, "filename"?: str}``.
    """
    if entity not in FILE_DOWNLOAD_ENTITIES:
        raise ValueError(f'File entity "{entity}" is not downloadable.')
    if not isinstance(id, str) or not _GUID_RE.fullmatch(id):
        raise ValueError(f"Invalid file id: {id}")
    if not BASE_URL:
        raise ValueError("Creatio base URL is not configured.")
    if not _bpmcsrf:
        await login()

    url = f"{BASE_URL}/0/rest/FileService/Download/{entity}/{id}"
    async with _client() as client:
        res = await client.get(url, headers=_session_headers("*/*"))
        if res.status_code in (401, 403):
            await login()
            res = await client.get(url, headers=_session_headers("*/*"))
    if res.status_code in (401, 403):
        raise AuthError(
            f"Creatio rejected the session (HTTP {res.status_code}) while downloading a file."
        )
    if not res.is_success:
        raise RuntimeError(
            f"File download {entity}/{id} -> {res.status_code} {res.reason_phrase}"
        )
    content_type = res.headers.get("content-type") or "application/octet-stream"
    cd = res.headers.get("content-disposition") or ""
    m = _FILENAME_RE.search(cd)
    out: dict[str, Any] = {"buffer": res.content, "contentType": content_type}
    if m:
        out["filename"] = m.group(1)
    return out


# ---------------------------------------------------------------------------
# OData query builder (read-only)
# ---------------------------------------------------------------------------
def _form_encode(s: str) -> str:
    """application/x-www-form-urlencoded, byte-for-byte like JS URLSearchParams.

    URLSearchParams leaves only ASCII alphanumerics and ``*-._`` unencoded and
    turns spaces into ``+``. quote_plus additionally keeps ``~``, so encode it.
    """
    return quote_plus(s, safe="*").replace("~", "%7E")


def _top_str(top: Any) -> str:
    # Mirrors JS String(n) feeding parseInt: 5.7 -> "5", missing -> default.
    if isinstance(top, bool):
        return ""
    if isinstance(top, float):
        return str(int(top)) if top == top and abs(top) != float("inf") else ""
    return str(top)


def build_query(opts: Mapping[str, Any] | None = None, **kw: Any) -> str:
    """Build ``?$select=...&$filter=...&$orderby=...&$expand=...&$top=N``.

    Options (dict or keywords): select (list[str]), filter, orderby, top, expand.
    ``$top`` is always present and clamped to 1..MAX_TOP.
    """
    o = {**(opts or {}), **kw}
    params: list[tuple[str, str]] = []
    if o.get("select"):
        params.append(("$select", ",".join(o["select"])))
    if o.get("filter"):
        params.append(("$filter", o["filter"]))
    if o.get("orderby"):
        params.append(("$orderby", o["orderby"]))
    if o.get("expand"):
        params.append(("$expand", o["expand"]))
    top = o.get("top")
    params.append(("$top", str(clamp_int(_top_str(MAX_TOP if top is None else top), MAX_TOP, 1, MAX_TOP))))
    s = "&".join(f"{_form_encode(k)}={_form_encode(v)}" for k, v in params)
    return f"?{s}" if s else ""


async def query_records(
    entity: str, opts: Mapping[str, Any] | None = None, **kw: Any
) -> list[Any]:
    """Convenience: query an entity set with the standard options. Returns the
    OData ``value`` array (or ``[]`` if absent)."""
    assert_entity_allowed(entity)
    data = await odata_get(entity + build_query(opts, **kw))
    return (data.get("value") if isinstance(data, dict) else None) or []


def connection_probe_path() -> str:
    """The OData path the connection check probes: one primary key from the
    first allowed entity. Public so the "always narrow $select" contract is
    testable without a network call.

    $select is deliberately narrowed to Id. An unqualified read asks Creatio to
    serialize EVERY column of the entity, so a single column that fails to
    serialize answers with "HTTP 500 ObjectContent`1 type failed to serialize"
    even on a perfectly valid session — which the Settings screen then reports
    as a credentials problem. One primary key tests exactly what this check is
    for: session validity and reachability.
    """
    entity = ALLOWED_ENTITIES[0] if ALLOWED_ENTITIES else "Contact"
    return entity + build_query(select=["Id"], top=1)


def _msg(e: BaseException) -> str:
    return str(e) or type(e).__name__


def describe_error(e: object) -> str:
    """httpx often reports transport failures with a terse message. Dig the
    real cause out of the exception chain (``__cause__`` / ``__context__``:
    connection resets, DNS failures, cert errors, connect timeouts) so the UI
    can show something actionable instead."""
    if not isinstance(e, BaseException):
        return str(e)
    top = _msg(e)
    parts: list[str] = []
    cur = e.__cause__ or e.__context__
    i = 0
    while cur is not None and i < 5:
        code = getattr(cur, "code", None)
        bit = code if isinstance(code, str) and code else _msg(cur)
        if bit and bit != top and bit not in parts:
            parts.append(bit)
        cur = cur.__cause__ or cur.__context__
        i += 1
    return f"{top} ({' → '.join(parts)})" if parts else top


async def test_connection() -> dict[str, Any]:
    """Lightweight auth/connectivity check — reads 1 row of the first allowed
    entity (or Contact). Returns ``{"ok": True}`` or ``{"ok": False, "error"}``."""
    try:
        await odata_get(connection_probe_path())
        return {"ok": True}
    except Exception as e:  # noqa: BLE001 — surfaced to the UI as text
        return {"ok": False, "error": describe_error(e)}


# pytest would otherwise collect this coroutine function as a test when a test
# module does `from creatio_case_lookup.creatio_client import *`-style imports.
test_connection.__test__ = False  # type: ignore[attr-defined]
